Framework/PythonInterface/mantid/fitfunctions.py

```diff
@@ -731,25 +755,32 @@ def _create_wrapper_function(name):
     """
     # ------------------------------------------------------------------------------------------------
     def wrapper_function(*args, **kwargs):
-        name_to_constructor = {
-            'CompositeFunction': CompositeFunctionWrapper,
-            'ProductFunction': ProductFunctionWrapper,
-            'Convolution': ConvolutionWrapper,
-            'MultiDomainFunction': MultiDomainFunctionWrapper,
-            }
-        # constructor is FunctionWrapper if the name is not in the registry.
-        if name in name_to_constructor:
-            return name_to_constructor[name](*args, **kwargs)
-        return FunctionWrapper(name, **kwargs)
+        return FunctionWrapper.wrap(name, *args, **kwargs)
 
     # ------------------------------------------------------------------------------------------------
     wrapper_function.__name__ = name
-    globals()[name] = wrapper_function
+    return wrapper_function
 
 
-def _create_wrappers_for_all_fit_functions():
+def _attach_wrappers(source_module):
     for name in FunctionFactory.getFunctionNames():
-        _create_wrapper_function(name)
+        if name not in _do_not_wrap:
+            setattr(source_module, name, _create_wrapper_function(name))
+
+
+_OldIFunction1D = IFunction1D
+
+
+class _NewIFunction1D(_OldIFunction1D):
+    """Overriden IFunction1D class that allows creation of
+    FunctionWrappers for newly defined fit functions.
+    """
+
+    def __new__(cls, _is_called_by_function_factory=False, **kwargs):
+        if _is_called_by_function_factory or (hasattr(cls.__init__, '__code__') and
```

Comment:
I think we could do with some comments here to explain what is going on and how the boolean connects back to the factory as the second part of the `if` is not completely clear here.